### PR TITLE
Reduced code duplication in ink_splash.dart, ink_ripple.dart

### LINKED ISSUE #21014: InkSplash and InkRipple have many pieces of identical code
We should combine those duplicate code pieces into a single one. See https://github.com/flutter/flutter/pull/20751#discussion_r211378022

### PR DESCRIPTION
**Pull request by mannprerak2**
## Description
Reduced code duplication in ink_splash and ink_ripple material widgets

## Related Issues
Fixes #21014

## Tests
Golden tests material.bottom_navigation_bar.shifting_transition.[0-7]

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by mannprerak2** [comment]
Continuing from the issue...
I would like to ask -
1. An appropriate name for this Mixin?
2. Where should I declare this mixin?
3. Should I also extract the following variables into the mixin?
```
  final Offset _position;
  final double _targetRadius;
  final TextDirection _textDirection;
```

File: packages/flutter/lib/src/material/ink_splash.dart [hunk 1]
```diff
@@ -75,6 +75,50 @@ class _InkSplashFactory extends InteractiveInkFeatureFactory {
   }
 }
 
+abstract class InkSplashRippleMixin {
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  // Reduces duplicate code between InkSplash and InkRipple
```
Comment (liyuqian) [inline packages/flutter/lib/src/material/ink_splash.dart]: Line 81 seems to fail the analyze test "ink_splash.dart:81: trailing U+0020 space character": https://cirrus-ci.com/task/5116804708171776?command=main#L8

File: packages/flutter/lib/src/material/ink_splash.dart [hunk 2]
```diff
@@ -75,6 +75,50 @@ class _InkSplashFactory extends InteractiveInkFeatureFactory {
   }
 }
 
+abstract class InkSplashRippleMixin {
```
Comment (liyuqian) [inline packages/flutter/lib/src/material/ink_splash.dart]: The comment in line 79, 80 will not be needed if we change `abstract class` to `mixin`

File: packages/flutter/lib/src/material/ink_splash.dart [hunk 3]
```diff
@@ -75,6 +75,50 @@ class _InkSplashFactory extends InteractiveInkFeatureFactory {
...
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  // Reduces duplicate code between InkSplash and InkRipple 
+  factory InkSplashRippleMixin._() => null;
+
+  /// paints a ripple/splash Ink Circle
+  void paintInkCircle(
```
Comment (liyuqian) [inline packages/flutter/lib/src/material/ink_splash.dart]: Since this mixin provides `paintInkCircle`, I'd recommend just name it as `InkCircleMixin`.

Even simpler, maybe we can just put this `paintInkCircle` as a `@protected` method inside `InteractiveInkFeature` as it's the parent class of both `InkRipple` and `InkSplash`. That will save us the trouble of coming up a name for the new class/mixin.

File: packages/flutter/lib/src/material/ink_splash.dart [hunk 4]
```diff
@@ -75,6 +75,50 @@ class _InkSplashFactory extends InteractiveInkFeatureFactory {
...
+    Canvas canvas,
+    Matrix4 transform,
+    Paint paint,
+    Offset center, {
+    @required TextDirection textDirection,
```
Comment (liyuqian) [inline packages/flutter/lib/src/material/ink_splash.dart]: If all these arguments are required, maybe just declare them as required positional arguments:
```
void paintInkCircle(
    Canvas canvas,
    Matrix4 transform,
    Paint paint,
    Offset center,
    TextDirection textDirection,
    Animation<double> radius,
    ShapeBorder customBorder,
    BorderRadius borderRadius,
    RectCallback clipCallback)
```

**Review by liyuqian — COMMENTED** [review]
> Continuing from the issue...
> I would like to ask -
> 
> 1. An appropriate name for this Mixin?

I think you can put `paintInkCircle` inside `InteractiveInkFeature` so you don't have to come up with a name.
> 2. Where should I declare this mixin?

As we can save the `mixin` by using `InteractiveInkFeature`, let's just put the common code inside `InteractiveInkFeature`'s file https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/ink_well.dart

Those above are all minor issues. My bigger concern is "Tests: none". @Piinks : can you confirm if there is any golden test for ink splash and ink ripple? If not, can you please guide @mannprerak2 on how to add a golden test for those two widgets?

**Review by Piinks — COMMENTED** [review]
I believe the `material.bottom_navigation_bar.shifting_transition.[0-7]` golden files capture the ink splash animation over multiple frames. There is also a `material.radio.ink_ripple` golden file test. All of these appear to be passing with this refactoring. :)
You can see the master images [here](https://flutter-gold.skia.org/list?fdiffmax=-1&fref=false&frgbamax=255&frgbamin=0&head=true&include=false&limit=50&master=false&match=name&metric=combined&neg=false&new_clstore=true&offset=0&pos=true&query=source_type%3Dflutter&sort=desc&unt=true).

**Comment by Piinks** [comment]
I believe @HansMuller would be a good authority here as well.

**Review by liyuqian — APPROVED** [review]
Thank you @Piinks for the reference! I updated the PR description to include the test info. LGTM and I think this PR is ready to merge if @HansMuller is Ok with it.

**Comment by HansMuller** [comment]
This looks like a good PR however it introduces a new bit of public (publicly visible) API, so it will need to be well documented and somewhat more defensive.

The overall purpose of the method needs a better explanation (it's not enough to just reiterate the method name) and all of the parameters need an explanation.

A short paragraph that mentions the classes which override this method would also be helpful.

Flutter usually uses named parameters for methods with this many arguments. Named parameters can be default null, with the (documented) default computed by the method. That seems to be the way the code has already been defined. Use @required parameters for parameters that must be specified, and assert what you can about the parameter values.

```dart
  @protected
  void paintInkCircle(
    Canvas canvas,
    Matrix4 transform,
    Paint paint,
    Offset center,
    TextDirection textDirection,
    Animation<double> radius,
    ShapeBorder customBorder,
    BorderRadius borderRadius,
    RectCallback clipCallback)
```

**Comment by liyuqian** [comment]
Thank you @HansMuller ! I didn't realize that `@protected` methods are also considered public APIs. @mannprerak2 : sorry for asking you to make the arguments positional. It seems that you'll have to change them back to named arguments.

**Comment by mannprerak2** [comment]
Pushed a commit with said changes...

File: packages/flutter/lib/src/material/ink_well.dart [hunk 5]
```diff
@@ -64,6 +64,54 @@ abstract class InteractiveInkFeature extends InkFeature {
     _color = value;
     controller.markNeedsPaint();
   }
+
+  /// Called during [paintFeature] in [InkSplash] and [InkRipple] widgets
+  ///
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: In between, as @HansMuller has recommended, there should be some documentation here on what the named parameters do since this function will be public. Assume that readers do not have any context to what this function is used for. Here's an example to get started:
```dart
/// If [customBorder] not null, the canvas is clipped by the [ShapeBorder] provided. If 
/// [customBorder] is null and [borderRadius] is non-null and not [BorderRadius.zero],
/// the canvas is clipped by the [BorderRadius] provided. {and so forth}
```

Comment (shihaohong) [reply packages/flutter/lib/src/material/ink_well.dart]: The rest of the parameters are still missing explanations/comments (canvas, transform, paint, center, etc).

Comment (mannprerak2) [reply packages/flutter/lib/src/material/ink_well.dart]: @shihaohong Added all remaining explanations except `canvas` as that seems redundant in case of painting

File: packages/flutter/lib/src/material/ink_well.dart [hunk 6]
```diff
@@ -64,6 +64,68 @@ abstract class InteractiveInkFeature extends InkFeature {
     _color = value;
     controller.markNeedsPaint();
   }
+
+  /// Draws an ink splash or ink ripple on the passed in [Canvas].
+  ///
+  /// The [transform] argument gives the coordinate conversion from the coordinate
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: I would reword this since [referenceBox] is unique to the implementation of ink splash and ink ripple. Maybe something more general like:
```
  /// The [transform] argument is the [Matrix4] transform that typically 
  /// shifts the coordinate space of the canvas to the space in which 
  /// the ink circle is to be painted. 
```

I'm not as familiar with the Canvas and painting operations, so maybe @HansMuller can chime in here.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 7]
```diff
@@ -64,6 +64,68 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// system of the canvas to the coordinate system of the [referenceBox].
+  ///
+  /// [Center], [Paint] and [radius] are used to paint the cirle with desired origin,
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: I'd break this apart into three different sections, since each argument can be further elaborated upon.

For Center, describe that it is the offset from the [transform]'s coordinate system and that it is where the center of the ink circle will be drawn.

For paint, describe interesting properties that might be useful or used in drawing the ink circle. 

The radius is pretty straightforward, but give it its own sentence.

For [center] and [paint], be sure to use lowercase "c" and "p" to ensure proper linking (for example, [Center] would likely link to the [Center] widget).

File: packages/flutter/lib/src/material/ink_well.dart [hunk 8]
```diff
@@ -64,6 +64,68 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// paint and radius
+  ///
+  /// [TextDirection] is used for calculating path to clip if [customBorder] is not null.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  /// [textDirection] is used for calculating path to clip if [customBorder] is not null.
```

At first glance, I'm unclear how [textDirection] is used by [customBorder] or what [customBorder] even is. Maybe start off with what [customBorder] is and then elaborate on how [textDirection] is used with [customBorder] here?

File: packages/flutter/lib/src/material/ink_well.dart [hunk 9]
```diff
@@ -64,6 +64,68 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///
+  /// If [customBorder] is not null, the canvas is clipped by the [ShapeBorder] provided. If
+  /// [customBorder] is null and [borderRadius] is provided (i.e not [BorderRadius.zero]),
+  /// the canvas is clipped by the [BorderRadius] provided.
+  ///
+  /// If [clipCallBack] is null, no clipping is performed and only the ink circle is painted
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: What is [clipCallback] and how is it used if it isn't null?

File: packages/flutter/lib/src/material/ink_well.dart (hunk 8 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/ink_well.dart]: Would also be a good place to describe the use of [borderRadius] here as well

File: packages/flutter/lib/src/material/ink_well.dart (hunk 6 above)
Comment (mannprerak2) [reply packages/flutter/lib/src/material/ink_well.dart]: I have changed this as you suggested as of now.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 10]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///
+  /// [center] is the [Offset] from origin of the canvas where center
+  /// of the circle is drawn
+  ///
+  /// Using [paint] you can specify properties like color, strokewidth, colorFiler, etc
+  /// for painting the circle
+  ///
+  /// [radius] is the radius of circle to be drawn on canvas
+  ///
+  /// [clipCallback] is required by ink effects to obtain the rectangle for the effect,
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: I think using the word required comes off too strong, since it implies that the parameter is required.
```suggestion
  /// [clipCallback] is the callback used to obtain the [Rect] used for clipping the ink effect.
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 11]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// if [clipCallback] is null, no clipping is performed and only the ink circle is painted
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  /// If [clipCallback] is null, no clipping is performed on the ink circle.
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 12]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the ink circle is to be painted.
+  ///
+  /// [center] is the [Offset] from origin of the canvas where center
+  /// of the circle is drawn
+  ///
+  /// Using [paint] you can specify properties like color, strokewidth, colorFiler, etc
+  /// for painting the circle
+  ///
+  /// [radius] is the radius of circle to be drawn on canvas
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  /// [radius] is the radius of ink circle to be drawn on canvas.
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 13]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// Draws an ink splash or ink ripple on the passed in [Canvas].
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the ink circle is to be painted.
+  ///
+  /// [center] is the [Offset] from origin of the canvas where center
+  /// of the circle is drawn
+  ///
+  /// Using [paint] you can specify properties like color, strokewidth, colorFiler, etc
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: - Helpful to link to the [Paint] object and be a little more specific with how it's used by the function.
- End sentences with period
```
  /// [paint] takes a [Paint] object that describes the style used to draw the ink circle.
  /// For example, [paint] can specify properties such as color, stroke width and 
  /// color filter.
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 14]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
   }
+
+  /// Draws an ink splash or ink ripple on the passed in [Canvas].
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the ink circle is to be painted.
+  ///
+  /// [center] is the [Offset] from origin of the canvas where center
+  /// of the circle is drawn
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: nit
```suggestion
  /// of the circle is drawn.
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 15]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///
+  /// Using [paint] you can specify properties like color, strokewidth, colorFiler, etc
+  /// for painting the circle
+  ///
+  /// [radius] is the radius of circle to be drawn on canvas
+  ///
+  /// [clipCallback] is required by ink effects to obtain the rectangle for the effect,
+  /// if [clipCallback] is null, no clipping is performed and only the ink circle is painted
+  ///
+  /// [customBorder] is a [ShapeBorder] which can be used if you want to clip
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: - Tailor the doc to be specific to the function (painting ink circles)
- End sentence with period
```suggestion
  /// [customBorder] is a [ShapeBorder] which can be used to clip the 
  /// ink circle with a custom path other than a [Rect] or [RRect]. 
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 16]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// something other than a simply a [Rect] or [RRect]
+  ///
+  /// If [customBorder] is provided, the canvas is clipped by it. If
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: I would be a little more granular here with what actually happens here:
```
  /// If [customBorder] is provided, [borderRadius] is ignored and the ink circle is 
  /// clipped by the [customBorder]'s outer path on the [Rect] provided by 
  /// [clipCallback]. 
  ///
  /// If [customBorder] is null and [borderRadius] is not [BorderRadius.zero], 
  /// the ink circle is clipped with rounded corners on the [Rect] provided by
  /// [clipCallback].
  /// 
  /// If [customBorder] is null and the [borderRadius] is [BorderRadius.zero],
  /// the ink circle will be clipped by the [Rect] provided by [clipCallback].
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 17]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// [customBorder] is null, canvas is clipped by the [RRect]
+  /// formed by [borderRadius] and [clipCallBack]
+  ///
+  /// [textDirection] is used if [customBorder] is provided (as [customBorder] can have
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: Elaborate on what a "[TextDirection] dependency" is:
```
/// [textDirection] is used by [customBorder] if it is non-null. This allows the [customBorder]'s path
/// to be properly defined if it was the path was expressed in terms of "start" and "end" instead of
/// "left" and "right".
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 18]
```diff
@@ -64,6 +64,79 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  /// a [textDirection] dependency), it is required for calculation of path to clip
+  ///
+  /// Used by [InkSplash] and [InkRipple], for example.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: - Prefer full sentences
```suggestion
  /// For examples on how the function is used, see [InkSplash] and [InkRipple].
```

**Comment by mannprerak2** [comment]
@shihaohong I've added a more complete explanation for clipping
```dart
  /// Clipping can happen in 3 different ways -
  ///  1. If [customBorder] is provided, it is used to determine the path
  ///     for clipping.
  ///  2. If [customBorder] is null, and [borderRadius] is provided, canvas
  ///     is clipped by a [RRect] created from [clipCallback] and [borderRadius].
  ///  3. if [borderRadius] is the default [BorderRadius.zero]. The [Rect] provided
  ///      by [clipCallback] is used for clipping.
```
and also made TextDirection non-required.. with added assertion to check that
TextDirection is provided if customBorder is provided.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 19]
```diff
@@ -64,6 +64,81 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///
+  /// [radius] is the radius of ink circle to be drawn on canvas.
+  ///
+  /// [clipCallback] is the callback used to obtain the [Rect] used for clipping the ink effect.
+  /// If [clipCallback] is null, no clipping is performed on the ink circle.
+  ///
+  /// Clipping can happen in 3 different ways -
+  ///  1. If [customBorder] is provided, it is used to determine the path
+  ///     for clipping.
+  ///  2. If [customBorder] is null, and [borderRadius] is provided, canvas
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  ///  2. If [customBorder] is null, and [borderRadius] is provided, the canvas
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 20]
```diff
@@ -64,6 +64,81 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///     is clipped by a [RRect] created from [clipCallback] and [borderRadius].
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  ///     is clipped by an [RRect] created from [clipCallback] and [borderRadius].
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 21]
```diff
@@ -64,6 +64,81 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///  3. if [borderRadius] is the default [BorderRadius.zero]. The [Rect] provided
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]:
```suggestion
  ///  3. If [borderRadius] is the default [BorderRadius.zero], then the [Rect] provided
```

File: packages/flutter/lib/src/material/ink_well.dart [hunk 22]
```diff
@@ -64,6 +64,81 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+    TextDirection textDirection,
+    ShapeBorder customBorder,
+    BorderRadius borderRadius = BorderRadius.zero,
+    RectCallback clipCallback,
+    }) {
+    assert(canvas != null);
+    assert(transform != null);
+    assert(paint != null);
+    assert(center != null);
+    assert(customBorder == null || textDirection != null, 'Must provide textDirection if customBorder is not null');
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/ink_well.dart]: I don't think we should assert that textDirection is non-null, since I don't believe `customBorder` will always have a `textDirection` dependency. The docs for ShapeBorder.outerPath say "The textDirection argument must be provided non-null _if_ the border has a text direction dependency"

**Review by shihaohong — COMMENTED** [review]
This is looking pretty close! Just the API doc comments and my note on removing the assert for textDirection.

**Comment by shihaohong** [comment]
You might also need to rebase your branch with `upstream/master`, it seems like some presubmit checks are failing for reasons outside of your PR

File: packages/flutter/lib/src/material/ink_well.dart [hunk 23]
```diff
@@ -64,6 +64,81 @@ abstract class InteractiveInkFeature extends InkFeature {
...
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the ink circle is to be painted.
+  ///
+  /// [center] is the [Offset] from origin of the canvas where center
+  /// of the circle is drawn.
+  ///
+  /// [paint] takes a [Paint] object that describes the styles used to draw the ink circle.
+  /// For example, [paint] can specify properties like color, strokewidth, colorFiler.
```
[the remark this answers is not in the document]
Comment (mannprerak2) [reply packages/flutter/lib/src/material/ink_well.dart]: Oops.., really sorry about this typo

File: packages/flutter/lib/src/material/ink_well.dart (hunk 22 above)
Comment (mannprerak2) [reply packages/flutter/lib/src/material/ink_well.dart]: Alright, I have removed it... Should I add anything to the dartdoc as well?

**Review by shihaohong — APPROVED** [review]
LGTM! Thanks for working on this

File: packages/flutter/lib/src/material/ink_well.dart (hunk 22 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/ink_well.dart]: I think it's clear enough
